tket/src/Clifford/ChoiMixTableau.cpp

```diff
@@ -0,0 +1,728 @@
+  }
+  return n;
+}
+
+unsigned ChoiMixTableau::get_n_outputs() const {
+  unsigned n = 0;
+  BOOST_FOREACH (
+      tableau_col_index_t::left_const_reference entry, col_index_.left) {
+    if (entry.first.second == TableauSegment::Output) ++n;
+  }
+  return n;
+}
+
+ChoiMixTableau::row_tensor_t ChoiMixTableau::stab_to_row_tensor(
+    const PauliStabiliser& stab) const {
+  QubitPauliMap in_qpm, out_qpm;
+  for (unsigned i = 0; i < stab.string.size(); ++i) {
+    Pauli p = stab.string.at(i);
+    col_key_t col = col_index_.right.at(i);
+    if (p != Pauli::I) {
+      if (col.second == TableauSegment::Input)
+        in_qpm.insert({col.first, p});
+      else
+        out_qpm.insert({col.first, p});
+    }
+  }
+  return {
+      QubitPauliTensor(in_qpm),
+      QubitPauliTensor(out_qpm, stab.coeff ? 1. : -1.)};
+}
+
+PauliStabiliser ChoiMixTableau::row_tensor_to_stab(
+    const row_tensor_t& ten) const {
+  std::vector<Pauli> ps;
+  for (unsigned i = 0; i < col_index_.size(); ++i) {
+    col_key_t qb = col_index_.right.at(i);
+    if (qb.second == TableauSegment::Input)
+      ps.push_back(ten.first.string.get(qb.first));
+    else
+      ps.push_back(ten.second.string.get(qb.first));
+  }
+  return PauliStabiliser(
+      ps, std::abs(ten.first.coeff * ten.second.coeff - 1.) < EPS);
+}
+
+ChoiMixTableau::row_tensor_t ChoiMixTableau::get_row(unsigned i) const {
+  return stab_to_row_tensor(tab_.get_pauli(i));
+}
+
+ChoiMixTableau::row_tensor_t ChoiMixTableau::get_row_product(
+    const std::vector<unsigned>& rows) const {
+  row_tensor_t result = {{}, {}};
+  for (unsigned i : rows) {
+    row_tensor_t row_i = get_row(i);
+    result.first = result.first * row_i.first;
+    result.second = result.second * row_i.second;
+  }
+  result.second.coeff *= result.first.coeff;
+  result.first.coeff = 1.;
+  return result;
+}
+
+void ChoiMixTableau::apply_S(const Qubit& qb, TableauSegment seg) {
+  unsigned col = col_index_.left.at(col_key_t{qb, seg});
+  tab_.apply_S(col);
+}
+
+void ChoiMixTableau::apply_V(const Qubit& qb, TableauSegment seg) {
+  unsigned col = col_index_.left.at(col_key_t{qb, seg});
+  tab_.apply_V(col);
+}
+
+void ChoiMixTableau::apply_CX(
+    const Qubit& control, const Qubit& target, TableauSegment seg) {
+  unsigned uc = col_index_.left.at(col_key_t{control, seg});
+  unsigned ut = col_index_.left.at(col_key_t{target, seg});
+  tab_.apply_CX(uc, ut);
+}
+
+void ChoiMixTableau::apply_gate(
+    OpType type, const qubit_vector_t& qbs, TableauSegment seg) {
+  switch (type) {
+    case OpType::Z: {
+      apply_S(qbs.at(0), seg);
+      apply_S(qbs.at(0), seg);
+      break;
+    }
+    case OpType::X: {
+      apply_V(qbs.at(0), seg);
+      apply_V(qbs.at(0), seg);
+      break;
+    }
+    case OpType::Y: {
+      apply_S(qbs.at(0), seg);
+      apply_S(qbs.at(0), seg);
+      apply_V(qbs.at(0), seg);
+      apply_V(qbs.at(0), seg);
+      break;
+    }
+    case OpType::Phase:
```

Comment:
The `Phase` operator is just a global phase; it is not the same as the `S` gate (yeah I just realized the name is confusing, sorry).